scripts/perf.js

```diff
@@ -0,0 +1,87 @@
+const puppeteer = require('puppeteer');
+const { performance } = require('perf_hooks');
+
+const SERVER = 'localhost';
+const PORT = 5000;
+const APP = 'test/perf/debug';
+
+async function createBrowser() {
+  const browser = await puppeteer.launch();
+
+  return {
+    openPage: async url => {
+      const page = await browser.newPage();
+      await page.goto(url);
+
+      return {
+        close: async () => page.close(),
+      };
+    },
+    close: async () => browser.close(),
+  };
+}
+
+async function runMeasures(
+  browser,
+  testCase,
+  times,
+) {
+  const measures = [];
+
+  for (let i = 0; i < times; i++) {
+    var start = performance.now();
+    const page = await browser.openPage(`http://${SERVER}:${PORT}/${APP}?${testCase}`);
+    const end = performance.now();
```

Comment:
Is there a way we can benchmark what's happening in the browser and send that back? Otherwise we'll add a lot of noise from what the browser is doing that is unrelated to the actual benchmark.

Seems like there's plenty of existing APIs for that: https://addyosmani.com/blog/puppeteer-recipes/#runtime-perf-metrics

Reply:
This looks much better, let me refactor to use the `metrics()`

Reply:
It might also be interesting to compare these `metrics()` to [React's Profiler component](https://reactjs.org/docs/profiler.html) (see also [How to use profiling in production](https://gist.github.com/bvaughn/25e6233aeb1b4f0cdb8d8366e54a3977)).

Reply:
Refactored using
```
    const perf = await page.evaluate(_ => {
      const { loadEventEnd, navigationStart } = performance.timing;
      return loadEventEnd - navigationStart;
    });
```

Reply:
Let me compare the numbers now with the React's Profiler component :)

Reply:
I've added the Profiler on the `box-emotion` example. Profiler result:

```
actualDuration: 19.274999969638884
baseDuration: 15.385000151582062
commitTime: 119.98000007588416
id: "box-emotion"
interactions: Set(0) {}
phase: "mount"
startTime: 98.64500002004206
```

perf result (10 different runs; this includes the <Profiler /> component as well):

```
Box emotion:

69.00ms
68.00ms
90.00ms
78.00ms
44.00ms
44.00ms
57.00ms
56.00ms
56.00ms
56.00ms
-------------
Avg: 61.80ms
```

Let me try this on more examples to see if the relative differences are similar.

Reply:
Here are more numbers:
1. box-emotion

```
Box emotion:

69.00ms
67.00ms
77.00ms
93.00ms
43.00ms
57.00ms
56.00ms
56.00ms
57.00ms
43.00ms
-------------
Avg: 61.80ms
```

```
actualDuration: 18.609998980537057
baseDuration: 15.669998829253018
commitTime: 110.84500001743436
id: "box-emotion"
interactions: Set(0) {}
phase: "mount"
startTime: 90.1650000596419
```

2. box-styled-components

```
Box styled-components:

67.00ms
63.00ms
64.00ms
80.00ms
97.00ms
61.00ms
61.00ms
47.00ms
48.00ms
48.00ms
-------------
Avg: 63.60ms
```

```
actualDuration: 15.089999069459736
baseDuration: 12.439999729394913
commitTime: 96.09999996609986
id: "naked-styled-components"
interactions: Set(0) {}
phase: "mount"
startTime: 79.29499994497746
```

3. box @material-ui/styles

```
Box @material-ui/styles:

123.00ms
122.00ms
89.00ms
103.00ms
103.00ms
85.00ms
121.00ms
123.00ms
84.00ms
84.00ms
-------------
Avg: 103.70ms
```

```
actualDuration: 44.76000100839883
baseDuration: 41.65500064846128
commitTime: 130.44999993871897
id: "box-material-ui-system"
interactions: Set(0) {}
phase: "mount"
startTime: 83.49999994970858
```

4. naked-styled-components

```
Box styled-components:

67.00ms
63.00ms
64.00ms
80.00ms
97.00ms
61.00ms
61.00ms
47.00ms
48.00ms
48.00ms
-------------
Avg: 63.60ms
```

```
actualDuration: 15.089999069459736
baseDuration: 12.439999729394913
commitTime: 96.09999996609986
id: "naked-styled-components"
interactions: Set(0) {}
phase: "mount"
startTime: 79.29499994497746
```

I am not sure how to compare the numbers of both as they are quite different... But I can conclude from both measurements for example that @material-ui/styles' Box is slower than the rest, not much else... We can try to render more instances of the components too, currently I am rendering 100 boxes. @eps1lon what do you think?

Reply:
I'd keep both: Browser metrics and React metrics.

Reply:
Added with https://github.com/mui-org/material-ui/pull/22923/commits/02fc972c0ac19f950b06280ac31fc3819216962d